vshard/service_info.lua

```diff
@@ -0,0 +1,181 @@
+local fiber_clock = require('fiber').clock
+local gsc = util.generate_self_checker
+
+--
+-- Helper function to get the number of seconds, minutes or hours
+-- that past since the time.
+--
+local function time_past_since(time)
+    local time_past = math.floor(fiber_clock() - time)
+    local units = {'s', 'm', 'h'}
+    for i, unit in ipairs(units) do
+        if time_past < 60 or not next(units, i) then
+            return time_past, unit
+        end
+        time_past = math.floor(time_past / 60)
+    end
+end
+
+--------------------------------------------------------------------------------
+-- Service description
+--------------------------------------------------------------------------------
+
+local SERVICE_TEMPLATE = {
+    -- Intended to show whether everything is all right.
+    status = 'ok',
+    -- Increasing number of status changes.
+    status_idx = 0,
+    -- Intended to show what the service is doing right now.
+    activity = 'unknown',
+    -- List of errors that have occurred and have to be fixed.
+    errors = rlist.new(),
+}
+
+local function service_set_status(service, status)
+    service.status_idx = service.status_idx + 1
+    service.status = status
+end
+
+-- Just for consistency
+local function service_set_activity(service, activity)
+    service.activity = activity
+end
+
+--
+-- Saves an error if the previous was not the same one.
+--
+local function service_add_error(service, err)
+    local errors = service.errors
+    -- Drop the oldest error if needed
+    if errors.count >= consts.DEFAULT_SERVICE_INFO_ERRORS_NUMBER then
+        errors:remove(errors.first)
+    end
+    -- Add a new one
+    if not service.errors.last or service.errors.last.error ~= err then
+        errors:add_tail({
+            error = err,
+            time = fiber_clock(),
+        })
+    end
+end
+
+local function service_set_status_ok(service)
+    service_set_status(service, 'ok')
+    -- Drop all errors
+    service.errors = rlist.new()
+end
+
+local function service_set_status_error(service, err)
+    -- 'status_idx' is updated even if no error will be saved
+    service_set_status(service, 'error')
+    service_add_error(service, err)
+end
+
+local service_mt = {
+    __index = {
+        -- Low-level setters
+        set_status = service_set_status,
+        set_activity = service_set_activity,
+        add_error = service_add_error,
+        -- High-level wrappers
+        set_status_ok = service_set_status_ok,
+        set_status_error = service_set_status_error,
+    }
+}
+
+--
+-- Wrap self methods with a sanity checker.
+--
+local index = {}
+for name, func in pairs(service_mt.__index) do
+    index[name] = gsc("service", name, service_mt, func)
+end
+service_mt.__index = index
+
+--------------------------------------------------------------------------------
+-- Module description
+--------------------------------------------------------------------------------
+
+local function module_add_service(module, name)
```

Comment:
I would advice to store the services individually, separated, not in a single table containing all the services. It just complicates access to them individually. So storage's `M` would have separate members for GC, rebalancer, recovery services. Each router would have separate bucket discovery and master location service members.

By storing the services in one table like now you probably only simplify the info() method a bit, but it is not worth the complication of having one another entity type like this `module` thing. You can just do `M.rebalancer = service_info.new()` and same for other services. It would do the same as now, but just without the `module` thing. Same for `module_info()` - you can just add `info` method to each service and get the info as `M.rebalancer:info()`. Try and you should see the simplification.

Reply:
Fixed